GPII-830 Onboard Zoomtext - Take 2

sgithens opened the pull request:
Keeping @JavierJF 's original PR open for now since it has comments and information on it:

https://github.com/GPII/universal/pull/735

sgithens commented:
@javihernandez @stegru @cindyli @amb26 I think this should be ready for review again. It looks like JJ had responded to the last items on the thread. I've updated this to master and the tests seem to have passed.

kaspermarkus commented:
@sgithens taking the liberty to mothball the original pull request and remove the morphic 1.3.5 and morphic 1.4 labels from it ... let me know if you disagree with that

sgithens commented:
That sounds fine @kaspermarkus , just so it's around in case we need to look at the historical comments.

the-t-in-rtf commented on gpii/node_modules/gpii-ini-file/index.js:
```diff
@@ -564,7 +564,7 @@ gpii.iniFile.readFile = function (path, options) {
     var content;
     try {
-        content = fs.readFileSync(path, "utf8");
+        content = fs.readFileSync(path, (options && options.encoding) || "utf-8");
```
couldn't you use `fluid.get` here?

the-t-in-rtf replied:
I guess it doesn't save much, and the path isn't that deep.  Happy to leave it.

the-t-in-rtf commented on testData/preferences/samuel.json5:
```diff
@@ -0,0 +1,81 @@
+
+                        // [VocalizerExpressive-Spanish Speech - Screen Reader]
+                        // ==================================================================
+                        "VocalizerExpressive-Spanish Speech - Screen Reader.Speaker": "",
```
These keys are insane.  We should never let anyone store these, and should use user-facing aliases instead.

sgithens replied:
probably, I guess I'll just leave them for now, similarly to how we did with the JAWS local language ones.

the-t-in-rtf replied:
I'll have to work on these when I formalise how we handle "intra-application" settings.

the-t-in-rtf commented on testData/solutions/win32.json5:
```diff
@@ -1,4 +1,6243 @@
+                            "default": 1
+                        }
+                    },
+                    // Bitmask Setting
```
IMO we should not onboard bitmask settings unless we have a way to handle them, like a transform from a bunch of individual settings.  In this case we also have no idea about the safe values or default.  I would suggest removing any that we don't have defaults and values for.  If there are any with a smaller range of inputs, we can enumerate them as enums/enumLabels.

sgithens replied:
these have all been removed except for the one with a default, minimum, and maximum

the-t-in-rtf commented on testData/solutions/win32.json5:
```diff
@@ -1,4 +1,6243 @@
+                        "schema": {
+                            "title": "Color effect",
+                            "description": "Selects a effect to be used in screen colors.",
+                            "enum": [
```
See, this is what I mean, this is obviously an enumerated bitmask, all the values are powers of two.

sgithens replied:
this has been removed as part of cleaning up based on SR test WARNINGS

the-t-in-rtf commented on testData/solutions/win32.json5:
```diff
@@ -1,4 +1,6243 @@
+                            "title": "Focus locator transparency",
+                            "description": "Specifies the transparency of the focus locator. Range goes from 'Clear' to 'Solid'.",
+                            "type": "integer",
+                            "minimum": 31,
```
It seems really suspicious that 31 would be "clear" instead of 0, but fine?

the-t-in-rtf replied:
Any confirmation here?  This still seems a bit suspicious to me.

sgithens replied:
Ok, the minimum and maximum values you can set in the slider are indeed 31 and 255

<img width="867" alt="image" src="https://user-images.githubusercontent.com/361700/82691397-b0eeb600-9c12-11ea-8612-75f026d8a74f.png">

the-t-in-rtf commented on testData/solutions/win32.json5:
```diff
@@ -1,4 +1,6243 @@
+                            "default": 2
+                        }
+                    },
+                    // Bitmask setting
```
I'd suggest searching for the label and/or comment and just ditching all of these bitmask fields.

sgithens replied:
These have all been removed as part of my cleanup based on your improved WARNING's from the SR, expect for one that had a legit default, minimum, and maximum value, so I guess I'll leave it for now?

the-t-in-rtf replied:
Yes, I have been discussing leaving whatever we can that's not obviously dangerous.  I'm fine with seldom-used settings as long as they aren't foot cannons like the unbounded/defaultless ones.

the-t-in-rtf commented on testData/solutions/win32.json5:
```diff
@@ -1,4 +1,6243 @@
+                            "default": 1
+                        }
+                    },
+                    "EchoMouse.LineMode": {
```
IMO settings keys should be alphabetised.  I can take care of it in the LSR work downstream.

sgithens replied:
Awesome, glad that can happen as part of the automagic LSR Work.

the-t-in-rtf commented on testData/solutions/win32.json5:
```diff
@@ -1,4 +1,6243 @@
+                            "default": 9
+                        }
+                    },
+                    // VOICES SECTION
```
I really think we should get rid of the voice-specific tweaks, and only keep things that can be configured generally for all voices.

sgithens replied:
it probably makes sense, but for the moment I'm leaving them there similarly to how we did for the JAWS pull from last week.

the-t-in-rtf replied:
It's fine, I have thoughts about using the power of IoC to represent these more simply, like reuse the common settings for each language.  I can revisit in the LSR work.

the-t-in-rtf commented on testData/solutions/win32.json5:
```diff
@@ -1,4 +1,6243 @@
+                            "default": 0
+                        }
+                    },
+          
```
"marning" should be "margin".

sgithens replied:
ugh, updated

the-t-in-rtf commented on testData/solutions/win32.json5:
```diff
@@ -1,4 +1,6243 @@
+                            "default": 0
+                        }
+                    },
+          
```
Another "marning" typo, I'd suggest searching for stragglers as well.

sgithens replied:
updated all 4 marnings

the-t-in-rtf commented:
I went through the pull, just to summarise the highlights, IMO:

1. We should remove any "bitmask" fields for which we don't have defaults and a reasonable set of known values that can be represented as an `enum` with `enumLabels`.
2. We should not support voice-specific settings if we can help it. If we need to support them, we at least should do something to clean up the keys.

I also see a lot of stuff that will be much nicer in the LSR era, if you could leave a note next to at least one of the On/Off enums for me, like "TODO: This should be inherited rather than defined repeatedly", I'd appreciate it.

the-t-in-rtf commented on testData/solutions/win32.json5:
```diff
@@ -1,4 +1,6243 @@
+                "type": "gpii.settingsHandlers.INISettingsHandler",
+                "liveness": "manualRestart",
+                "options": {
+                    "filename": "${{environment}.AppData}\\Freedom Scientific\\ZoomText\\2019\\Config\\zten-US.zxc",
```
This is probably gonna need to be updated to 2020.

sgithens replied:
You would think so, but we should probably check with the pilot/case studies team on this...

the-t-in-rtf commented on tests/platform/windows/windows-zoomtext-testSpec.js:
```diff
@@ -0,0 +1,163 @@
+                            "Tracking.RouteMouse": 0
+                        },
+                        "options": {
+                            "filename": "${{environment}.AppData}\\Freedom Scientific\\ZoomText\\2019\\Config\\zten-US.zxc",
```
Search for `ZoomText\\2019` and replace with `ZoomText\\2020` throughout.

sgithens replied:
similar as above, should double check to see which version we actually plan on using this summer

sgithens commented:
@the-t-in-rtf I think I've addressed all the comments... expect I didn't remove the various voice specific settings from the various languages... similarly to how we ended up not removing them for JAWS. They are indeed ugly, but I don't think there is any better way to do them at the moment. Unless we just keep one or two of them and delete the other voices...

the-t-in-rtf commented:
If you can find an answer to my question about `31` being clear in that one setting, I'm happy with the reset.

sgithens commented:
@the-t-in-rtf Ok, I verified the minimum value above (with screenshot), and chatted with Amarja who is still indeed testing with 2019 at the moment, so let's just leave it at the current version year for now.
